addons/mail/models/res_users.py

```diff
@@ -35,12 +35,30 @@ class Users(models.Model):
         "Only internal user can receive notifications in Odoo",
     )]
 
-    @api.depends('share')
+    @api.depends('share', 'groups_id')
     def _compute_notification_type(self):
+        inbox_group_id = self.env['ir.model.data']._xmlid_to_res_id('mail.group_mail_notification_type_inbox')
+
+        self.filtered_domain([
+            ('groups_id', 'in', inbox_group_id), ('notification_type', '!=', 'inbox')
+        ]).notification_type = 'inbox'
+        self.filtered_domain([
+            ('groups_id', 'not in', inbox_group_id), ('notification_type', '=', 'inbox')
+        ]).notification_type = 'email'
+
+        # Special case: internal users with inbox notifications converted to portal must be converted to email users
+        self.filtered_domain([('share', '=', True), ('notification_type', '=', 'inbox')]).notification_type = 'email'
+        self.filtered_domain([('share', '=', True), ('groups_id', 'in', inbox_group_id)]).groups_id = [
+            Command.unlink(inbox_group_id)
+        ]
```

Comment:
This unlink will be done in the inverse method so it is not necessary to do that here, I think? 🤔

Reply:
Well, indeed. 

It's funny though: It is tested in the unit test `test_notification_type_convert_internal_inbox_to_portal` I included in this PR. To switch the user from internal to portal, I first used
```py
user.groups_id -= self.env.ref('base.group_user')
user.groups_id += self.env.ref('base.group_portal')
```
While using this above version, I had to add this `Command.unlink(inbox_group_id)` in the compute method to make the test work.
I then switched to
```py
user.write({'groups_id': [
    Command.unlink(self.env.ref('base.group_user').id),
    Command.link(self.env.ref('base.group_portal').id),
]})
```
And for this one this is no longer needed.

I just tested this extensively, using the `-=`, `+=` version is buggy, it doesn't remove automatically the implied groups as it should. For instance, `Mail / Template Editor` stays while using `-=`, `+=`, it doesn't when using the write version.

I removed this `Command.unlink(inbox_group_id)` from the compute.